add debugger class and connect to debugger websocket

grgustaf opened the pull request:
JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf committed:
add debugger class and connect to debugger websocket

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

grgustaf commented:
Got past the promise stuff and got it simply connected, but I'll continue working on this patch

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,47 @@
+
+  constructor(address: string) {
+    // NOTE: It seems pretty clean to keep the defaults hidden within the class like this.
+    // This was how the .py version did it, maybe I should do this for proxy too?
```
Hmm, I think there's utility in being able to `import` the defaults (in test scenarios is one use case for example).

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,47 @@
+  private _socket: WebSocket | undefined = undefined;
+  private _connectPromise: Promise<void>;
+
+  constructor(address: string) {
```
I'd instead pass an "options object" here and add a `interface ...Options { host: string; port: number; ... }` like you have in the other file.

This is a bit more flexible compared to passing arguments, esp. with an eye on future extensibility.

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,47 @@
+    this.host = hostAndPort[0] || 'localhost';
+    this.port = parseInt(hostAndPort[1], 10) || 5001;
+
+    this._connectPromise = new Promise((resolve, reject) => {
```
The executor function will be run "immediately", so you're immediately creating this `WebSocket` connection. I wonder if it's more desirable to have dedicated `connect` and `disconnect` methods.

grgustaf replied:
I started out with a connect function and went back and forth in my mind. If you think that makes more sense, I'm fine with it. It just seemed kind of funny to not just connect, if we're already passing in the address - an extra step. But I guess it can be disconnected and connected again, it makes sense.

martijnthe replied:
Yeah, I think it would make sense to be able to connect, disconnect and reconnect the same instance.

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,47 @@
+
+      this._socket.on('error', () => {
+        reject();
+      });
```
Will `error` only happen upon a failed connection attempt? In other words, not later during the connection, say when sending a message fails?

grgustaf replied:
I was going back and forth about such things a lot myself. But the promise would have already been resolved by that point, so it won't change anything. I'm not sure whether cleaning up after such code is important? I considered doing removeListener() for example.

martijnthe replied:
Hmm yeah, I guess that's true.

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,47 @@
+  readonly host: string;
+  readonly port: number;
+  private _socket: WebSocket | undefined = undefined;
+  private _connectPromise: Promise<void>;
```
No need to `_` suffixing, `private` already indicates it.

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,47 @@
+
+  getConnectPromise(): Promise<void> {
+    return this._connectPromise;
+  }
```
I'd probably do something along these lines:
```
connect(): Promise<void> {
  if (!this.connectPromise) {
    this._connectPromise = new Promise((resolve, reject) => {
      // what you already have
    }
  }
  return this.connectPromise;
}
```

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,47 @@
+    this._connectPromise = new Promise((resolve, reject) => {
+      this._socket = new WebSocket(`ws://${this.host}:${this.port}/jerry-debugger`);
+      this._socket.binaryType = 'arraybuffer';
+
```
Also listen to `close` events and clear out the promise? `this._connectPromise = undefined;`

grgustaf replied:
By returning the promise from connect() I think I don't actually need to hold onto it anymore in the class.

martijnthe commented on jerry-debugger/src/cli/cli.ts:
```diff
@@ -45,17 +48,24 @@ export function getOptionsFromArgs(argv: Array<string>) {
+      port: options.port,
+      jsfile: options.jsfile,
+    });
+  }).catch(() => {
```
Maybe grab the error and print it too?

`catch((e) => { ... // do something with e }`

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,47 @@
+export class JerryDebugger {
+  readonly host: string;
+  readonly port: number;
+  private _socket: WebSocket | undefined = undefined;
```
Could write `_socket?` instead of `... | undefined`

grgustaf replied:
Ah, thanks.

martijnthe commented on jerry-debugger/src/cli/cli.ts:
```diff
@@ -45,17 +48,24 @@ export function getOptionsFromArgs(argv: Array<string>) {
     verbose: args['verbose'],
-    jsfile: args._[0],
+    address: args._[0] || '',
+    jsfile: args['client-source'] || 'untitled.js',
```
I'd like to stick with the node.js CLI semantics as much as possible because this will make integrating with existing IDEs/editors much easier:

```
$ node --help
Usage: node [options] [ -e script | script.js | - ] [arguments]
       node inspect script.js [arguments]

Options:
  -v, --version              print Node.js version
  -e, --eval script          evaluate script
  -p, --print                evaluate script and print result
  -c, --check                syntax check script without executing
  -i, --interactive          always enter the REPL even if stdin
                             does not appear to be a terminal
  -r, --require              module to preload (option can be repeated)
  -                          script read from stdin (default; interactive mode if a tty)
  --inspect[=[host:]port]    activate inspector on host:port
                             (default: 127.0.0.1:9229)
  --inspect-brk[=[host:]port]
                             activate inspector on host:port
                             and break at start of user script
  --inspect-port=[host:]port
                             set host:port for inspector
... etc ...
```

martijnthe replied:
For new, JerryScript specific arugments, let's prefix with `--jerry-...` to avoid conflicts. How does that sound? 

So perhaps `--jerry-remote[=[host:]port]`?

(Side note: I also think we should try to think beyond using a network host:port as the way to connect. I can imagine people using file socket, i.e. `/dev/serial.tty` or perhaps using a USB profile or even Bluetooth. I think we can bolt those opto `--jerry-remote` later on i.e. `--jerry-remote=usb://....`)

grgustaf replied:
Good idea... I was trying to preserve the existing JrS command line syntax but this makes more sense.

grgustaf committed:
add connect/disconnect functions to JerryDebugger

Also, reuse code for parsing host/port, handle options consistently,
rename command line parameters in line with node.

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,66 @@
+
+  connect() {
+    if (this.socket) {
+      this.disconnect();
```
Why disconnect if already connected?

Perhaps `return Promise.resolve();`?

grgustaf replied:
This does create a (very?) minor bug though, because if connect() is called before the socket has finished connecting the resolving promise indicates it has connected. Maybe better to throw as it should never happen anyway?

martijnthe replied:
Oh, shoot I only saw this after (squash-)merging.

Good catch. Better to this assign the already created promise to an instance variable and return that instead of creating a new promise.

martijnthe commented on jerry-debugger/src/lib/debugger-client.ts:
```diff
@@ -0,0 +1,66 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import WebSocket from 'ws';
```
JFYI, when writing unit tests for this code, you can use `jest.mock('ws')` to let Jest auto-mock the module.

martijnthe commented:
@grgustaf jfyi, I posted a couple more comments on this one.

grgustaf committed:
don't reconnect, leave existing connection open

JerryScript-DCO-1.0-Signed-off-by: Geoff Gustafson geoff@linux.intel.com
